Improve PathSimplification Performance and Memory Overhead

boldtrn opened the pull request:
As reported in #1162, there are issues with the performance and memory usage.

One issue of memory usage was found in the PathList#copy method. 

I profiled the current master and found that even with 3051eca5683ed9bf1446419fd9e0c4052c33a5a5 a lot of memory is used due to PathList#copy and PathList#add. This PR tries to solve this to solve this issue by creating [shallow copies](https://en.wikipedia.org/wiki/Object_copying#Shallow_copy) of the PathList instead of deep copies. I cannot find the same memory consumption with this PR.

The issue with the shallow copies is that we might see incorrect results if we change the original PointList. For example we have a PointList with 3 points, A,B,C and we create a Shallow copy of this PointList, referencing point A, B. If we now change the underlying PointList, for example reverse it, then the shallow copy references to the points C,B instead of A,B.

We cannot simply lock the PointList, once we create a shallow copy, since we create shallow copies during the PathSimplification after each simplification (see [PathSimplification#reduceLengh](https://github.com/graphhopper/graphhopper/blob/20d6e67bf114543ea6feaabf5510510a14eecc8d/core/src/main/java/com/graphhopper/util/PathSimplification.java#L160), but change the original PointList with each simplification. It might be possible to lock the original PointList once we merge #1138. A partial lock might work, but I am not sure if this becomes too much of an overhead?

I benchmarked the performance of this PR with Apache Bench and the following command: `ab -n 1000 -c 20 "http://localhost:8989/route?point=49.214009%2C8.723145&point=49.203243%2C10.151367&point=47.755945%2C9.695435&point=48.991032%2C9.613037&point=48.026672%2C7.973328&type=json&locale=en-AU&vehicle=car&weighting=fastest&elevation=false&key="`

So it sampled 1000 requests and 20 concurrent requests. The used route is about 750km long.

Master:
```
Concurrency Level:      20
Time taken for tests:   14.157 seconds
Complete requests:      1000
Failed requests:        6
   (Connect: 0, Receive: 0, Length: 6, Exceptions: 0)
Total transferred:      55914988 bytes
HTML transferred:       55505994 bytes
Requests per second:    70.64 [#/sec] (mean)
Time per request:       283.139 [ms] (mean)
Time per request:       14.157 [ms] (mean, across all concurrent requests)
Transfer rate:          3857.08 [Kbytes/sec] received
```

This PR:
```
Concurrency Level:      20
Time taken for tests:   13.346 seconds
Complete requests:      1000
Failed requests:        12
   (Connect: 0, Receive: 0, Length: 12, Exceptions: 0)
Total transferred:      55914976 bytes
HTML transferred:       55505988 bytes
Requests per second:    74.93 [#/sec] (mean)
Time per request:       266.920 [ms] (mean)
Time per request:       13.346 [ms] (mean, across all concurrent requests)
Transfer rate:          4091.46 [Kbytes/sec] received
```

For some reason `./graphhopper.sh measurement` does not produce reproducible results for me. I have not found the reason for this but it might differ +-5% per measurement. So sometimes it seemed this PR would improve the performance, sometimes it seems it does not. With Apache Bench it seems to improve the performance.

karussell commented:
> The issue with the shallow copies is that we might see incorrect results if we change the original PointList. 

This is ugly. But unavoidable IMO. To make it a bit more safe it might be interesting to have an explicit makeReadOnly method after the simplification? So while simplification there is some risk for our internal development but afterwards everything is fine for the end user? To make this more obvious we could rename the method to shallowUnsafeCopy?

karussell commented:
I can confirm that this improves the introduced instructions overhead reported in #1162.

But I can use Measurement. Also I found that the slow down was introduced in commit 

11efb7ae6289a66b7a34644b2b40e12bb49de40f Add Average Speed as PathDetails to Response (#1091)

~3.6ms vs. 4.1ms

Now with this change it is back to 3.7ms but of course only if I delete the newly introduced lines in Measurement:

```java
if (withInstructions)
    req.setPathDetails(Arrays.asList(Parameters.DETAILS.AVERAGE_SPEED));
```

boldtrn commented:
> if (withInstructions)
    req.setPathDetails(Arrays.asList(Parameters.DETAILS.AVERAGE_SPEED));

Ah, this explains a lot :). I forgot that we add PathDetails in der Measurement and didn't change that line.

> To make it a bit more safe it might be interesting to have an explicit makeReadOnly method after the simplification?

Yes I think a write lock would be a good addition here. Will add this.

> To make this more obvious we could rename the method to shallowUnsafeCopy?

Yes, makes sense.

boldtrn commented on core/src/main/java/com/graphhopper/util/PathMerger.java:
```diff
@@ -142,6 +142,8 @@ public void doWork(PathWrapper altRsp, List<Path> paths, Translation tr) {
             PathSimplification ps = new PathSimplification(altRsp, douglasPeucker, enableInstructions);
             ps.simplify();
+            // Make sure that the instruction references are not broken
+            altRsp.getPoints().makeImmutable();
```
I wasn't sure if it would be better to put this in the simplification, but then decided it makes more sense here, so it's not hidden and it's more obvious to others what happens here.

karussell replied:
Good question :)

I would tend to put it in PathSimplication as then all uses of it are already safe. 

BTW: what happens if we have an immutable PointList and pass it to PathSimplification - will this fail?

boldtrn replied:
Ok, I can change this. 

> BTW: what happens if we have an immutable PointList and pass it to PathSimplification - will this fail?

Haven't tried it, but most probably, yes. Right now I haven't added a way to unlock a PointList. Once Locked it stays locked. You will have to create a deep copy to have the same PointList unlocked, which kind of makes sense to me, as we promised that the safeShallowCopy would be actually safe.

boldtrn commented on core/src/test/java/com/graphhopper/util/PointListTest.java:
```diff
@@ -110,4 +111,46 @@ public void testCopy_issue1166() {
+        }
+
+        PointList pl4 = pl1.shallowCopyUnsafe(0, pl1.size());
+        assertTrue(pl1.equals(pl4));
```
Is this correct? Yes it contains the same points, but it is not really equal as one is only a shallow copy.

karussell replied:
I would say, yes.

boldtrn commented:
Ok, at first I implemented the `equals` function, that it would also check if the mutability of both PointLists is equal. This made some tests fail, as assertEquals failed. For now I reverted the change but I am not sure, if equals should check for the mutability of the object as well?

karussell commented on core/src/main/java/com/graphhopper/util/PointList.java:
```diff
@@ -138,12 +138,13 @@ public boolean is3D() {
     };
     private final static DistanceCalc3D distCalc3D = Helper.DIST_3D;
-    private static String ERR_MSG = "Tried to access PointList with too big index!";
+    protected static String ERR_MSG = "Tried to access PointList with too big index!";
```
just `final static` is better here. `protected` is too open

karussell commented on core/src/main/java/com/graphhopper/util/PointList.java:
```diff
@@ -216,6 +220,9 @@ public void add(double lat, double lon) {
     }
 
     public void add(double lat, double lon, double ele) {
+        if (this.isImmutable()) {
```
To reduce redundancy I would move this into a ensureImmutability method

boldtrn replied:
Ah yes, makes sense, thanks :)

boldtrn replied:
Thanks, I name the method `ensureMutability` as we are making sure that the object is mutable, I think it is a bit clearer that way?

karussell commented on core/src/main/java/com/graphhopper/util/PointList.java:
```diff
@@ -449,14 +468,41 @@ public PointList copy(int from, int end) {
+     * <p>
+     * If you know what your are doing, you can also use {@link PointList#shallowCopyUnsafe(int, int)}.
+     */
+    public PointList shallowCopySafe(final int from, final int end) {
```
Do we need the safe method? Or maybe use unsafe as a parameter?

boldtrn replied:
Would be also possible. I liked the idea of two different methods here, but doesn't really matter and keeps the duplication low. I can merge the two methods.

karussell replied:
I do not like the name with safe+unsafe (not exactly clear to which word the safe belongs) and shallowCopy (or createShallowCopy?) is nicer

karussell commented on core/src/main/java/com/graphhopper/util/ShallowImmutablePointList.java:
```diff
@@ -0,0 +1,205 @@
+            }
+            sb.append(')');
+        }
+        return sb.toString();
```
Is this duplication necessary?

boldtrn replied:
I just spotted an issue here, thanks. The method should run from `from` to `end` and not from 0 to size. I can try to reduce the code duplicates a bit :).

karussell commented:
Thanks. One final safety improvement we should discuss: should we make both PointList classes final? Currently one could tweak the PointList endlessly which could introduce methods with which the shallow copy has problems etc

boldtrn commented:
> One final safety improvement we should discuss: should we make both PointList classes final? 

Would this work? The ShallowCopy extends the PointList. I think this would be forbidden if we set the PointList final?

karussell commented:
Ah, ok. That is still a bit ugly. Hmmh.

boldtrn commented:
> Ah, ok. That is still a bit ugly. Hmmh.

I added a hint in the class comment of the PointList. I think if someone tries to extend this class he/she would at least read the class comment. Probably there is not much we can do, if someone incorrectly extends any class of GH, it might break?

karussell commented:
> Probably there is not much we can do, if someone incorrectly extends any class of GH, it might break?

If some classes are not explicitly intended for this we should make them final (I know this is not the case elsewhere in the code base but we should improve step by step :)).

There is the following trick that we could try: create a _PointList class that is package protected and that does what the current PointList does, then create the two **final** classes: `PointList extends _PointList` and `ShallowCopyPointList extends _PointList`. Would this work?

boldtrn commented:
> If some classes are not explicitly intended for this we should make them final (I know this is not the case elsewhere in the code base but we should improve step by step :)).

But wouldn't it be better to offer users the flexibility to extend and change GraphHopper very flexible? Or what is the reason to forbid this explicitly?

> There is the following trick that we could try: create a _PointList class that is package protected and that does what the current PointList does, then create the two final classes: PointList extends _PointList and ShallowCopyPointList extends _PointList. Would this work?

It could work like this, yes. But this makes usages of the PointList class outside of the package ugly. If we would keep the code as it is, we could not simply use the ShallowCopy instead of the PointList and we cannot set the datatype to _PointList as this class is package protected. So we would have to set the PointList of the Instructions to ShallowImmutablePointList and change all interfaces they use together. So maybe we can trick Java to accept this by casting the ShallowCopy to a PointList. Personally, I would prefer to avoid this.

karussell commented:
> But this makes usages of the PointList class outside of the package ...

Ehem, yes ;) ... so this won't work. The only solution left would be to break the API and introduce a PointList interface, but this would be too ugly. Let us keep it this way and make at least the shallow class final.

>  But wouldn't it be better to offer users the flexibility to extend and change GraphHopper very flexible? Or what is the reason to forbid this explicitly?

We can open up our API if we know there really is a good use case. Otherwise if we allow broad usage we later cannot easily migrate to other stuff or implementation OR we have to break this public API.

boldtrn commented:
> Otherwise if we allow broad usage we later cannot easily migrate to other stuff or implementation OR we have to break this public API.

Ah, now I understand :). Ok, makes sense.

> Let us keep it this way and make at least the shallow class final.

Just pushed the change :).
